### DISCUSSION TITLE
kitty vi mode, like alacritty

### DISCUSSION THREAD

**Discussion by xenoterracide** [kovidgoyal/kitty#3904] [Q&A]
I dislike alacritty because it doesn't have tabs and wants me to use tmux for that feature even though it seems using tmux breaks the good things about it. Can I configure kitty to behave as a vi? so I can navigate things like the scrollback with vi keys and highlight in a visual mode, etc? and ideally switch tabs with :gt :gT ? I didn't see any of this in keyboard mappings but maybe there's something possible via a lot of custom ones?

**Comment by kovidgoyal** [comment]
https://sw.kovidgoyal.net/kitty/overview/#the-scrollback-buffer

**Reply by xenoterracide** [reply]
Not as nice as I was hoping for, it's serviceable, but less is... aging a bit, I actually looked for a replacement recently and didn't find one. Technically, I could use vim for that.

And of course less can't be used to switch the tabs back and forth. Would be cool to get all this as more of a built-in, where Alacritty is refusing to support tabs 🙄. I need one terminal that can do it all! ;) :D

**Reply by kovidgoyal** [reply]
That link shows you how to use vim as well. And I dont know what you mean by switching tabs back and forth, but kitty has those capabilities in spades, read the docs.

**Reply by xenoterracide** [reply]
I mean if Kitty could enter a command mode like it was vim itself. Then you could use vim keys to navigate Kitty without going into a pager. Vim has both the concept of tabs and windows so you could use the same key bindings as them uses to navigate tabs and windows as well as scrolling. With just dropping into them or less or whatever as a pager that doesn't allow you to use vim navigation keys to control kitty itself.

And then once you enter command mode if you have more than one tab :gt will go to the next tab, :gT is the previous tab. :tabn iirc is a new tab. 

Simply entering a pager doesn't get you this behavior. Trying out alacritty and then thinking well what if it could use them skis to switch tabs.... Or windows or both. Also trying out that behavior in vim itself, would probably show more what I'm talking about. 

It's just using vim command mode style to control Kitty instead of your normal shortcuts.

**Comment by kovidgoyal** [comment]
Why would you want to type more keys to achieve actions? And if you
really want to type commands tocontrol kitty, use the kitty shell
ctrl+shift+esc

**Reply by xenoterracide** [reply]
Because it's a learning curve and yes I can learn it but with every terminal being different and not being able to guarantee exactly what I'm going to encounter It's a lot more overhead on the brain. Yes it's more keystrokes but you know how to do it off the top of your head. VI is VI, and I've never seen a server without one. I mean obviously most of these terminal emulators that I'm talking about don't support this either, after trying alacrity I loved it, but they absolutely refuse to support tabs, and unlike alacrity you have the opinion that multiplexers are generally a bad idea.

I thought about forking alacrity and asking my ward to patch it for tabs. And maybe some of the other patches, you also currently support ligatures with alacrity does not support yet.


Vim does have a few options and I kind of hate that they've been replacing them with other ones as the default. If I wanted busy box vi I just because it's smaller, I would have installed that.

**Reply by kovidgoyal** [reply]
You want your terminal to act like an editor? How would that work? the two do very different things. Terminals are not modal for one thing. But if you want to you can recreate vim keybindings in kitty, for example:

map leader>g>t next_tab

where you choose some key combination for leader. You cant use : like you would in vim since that wold prevent you for typing colons, so use ctrl+; or whatever you like.

**Reply by xenoterracide** [reply]
Yeah it probably does tie in with that. I think you could always leave the file manually editable so that. People could change that if needed. I mean maybe there needs to be an option too not persisting it that way, in case you need other logic around it and your zshrc rather than just a simple list.

**Reply by NicTanghe** [reply]
the point of vim being a standard is that you don't need to set up custom things.
native implementation is a great idea.
Having every user set it up themselves is a shit solution.

**Comment by xenoterracide** [comment]
Again, you download Alacritty and give that a try and see how it works, it doesn't do tabs or windows, but the principle is the same. https://github.com/alacritty/alacritty besides, they're one of your biggest competitors, not that that matters in open source.

**Reply by kovidgoyal** [reply]
What principle? I have no idea what you are talking about. I find using an actual editor, one that I know how to use intimately and is fully customizable and *much* more powerful than just a set of keybindings, to navigate scrollback strictly superior to some kludged together hack like alacritty has. If you disagree feel free to use alacritty.  kitty is *not* going to have a poorly implemented partial vi mode to navigate scrollback, instead it lets you ACTUALLY USE VI to do it.

**Reply by Bestboigill** [reply]
> What principle? I have no idea what you are talking about. I find using an actual editor, one that I know how to use intimately and is fully customizable and _much_ more powerful than just a set of keybindings, to navigate scrollback strictly superior to some kludged together hack like alacritty has. If you disagree feel free to use alacritty. kitty is _not_ going to have a poorly implemented partial vi mode to navigate scrollback, instead it lets you ACTUALLY USE VI to do it.

@kovidgoyal Dude why are you so angry about it?

If it was as sense-less maybe other terminals would not have implemented it , even powershell has a vi mode 

Are you the actual dev to be so hurt about what it should and shouldn't have

**Reply by NicTanghe** [reply]
probably because it makes sheeple want to agree with him as you can see with the 18 likes.

**Comment by fladson** [comment]
I am sorry to put my nose on this but nvim terminal seems to be a better option for you.
You can also navigate the scrollback easily with vim, as @kovidgoyal said, I currently have this:

`scrollback_pager nvim -c 'setlocal ft=man nonumber nolist showtabline=0 foldcolumn=0' -c "autocmd VimEnter * normal G" -`

The color highlight is not perfect, but it gets the job done.

**Reply by JarnoB** [reply]
Thanks, this was exactly what I was looking for! 

I noticed it inserted a blank line for each line that was blank in the terminal as wel, so I fixed that:

`scrollback_pager nvim -c 'setlocal ft=man number modifiable noro relativenumber nolist showtabline=0 foldcolumn=0' -c "autocmd VimEnter * normal G" -c ":silent! %s#\($\n\s*\)\+\%$##" -
`  

`modifiable` enables vim to modify the text. However this gave me a warning that it was an read-only file, so I supressed that warning with `noro`. The part at the end removes the blank lines.

I also added number and relativenumber, but that's just personal preference.

Edit: Forgot to add `:silent!` before. You wan to add this, as it will otherwise give you an error-message if you don't have any blank lines.

**Comment by jclsn** [comment]
For me the output is inserted into vim with with all character codes

```[m[38:5:76m❯[39m [32mls```

How can I turn this off?

**Reply by page-down** [reply]
> For me the output is inserted into vim with with all character codes

You need to use the terminal feature in vim.
There are also some examples in issue #719.
Or feed plain text to vim.
https://sw.kovidgoyal.net/kitty/launch/#the-piping-environment

Here is my configuration using neovim as pager.
No temporary files are written and there are no additional processes.

https://github.com/kovidgoyal/kitty/issues/719#issuecomment-952039731

**Reply by jclsn** [reply]
Yes! Perfect! Thanks :)

**Reply by jclsn** [reply]
Ah no it is not perfect. Every solution I have tried so far is very buggy. 

@kovidgoyal What do you use? I also don't really see your point in how this is superior if no one has found a working solution so far. I mean, of course vim is superior to Alacritty's vi mode. I would also like to use all its commands if it worked realiably, but it so far it doesn't. While being in an ssh session the piping to vim only works by chance.

**Reply by lacygoill** [reply]
> Ah no it is not perfect. Every solution I have tried so far is very buggy.

FWIW, this is what I'm using in `~/.config/kitty/kitty.conf`:

    scrollback_pager bash -c 'vim --not-a-term +"terminal ++curwin cat "<(cat) +"KittyScrollback INPUT_LINE_NUMBER CURSOR_LINE CURSOR_COLUMN"'

And this is how `:KittyScrollback` is defined in my `~/.vim/vimrc`:

    vim9script

    command -bar -nargs=+ KittyScrollback KittyScrollback(<f-args>)

    def KittyScrollback(topline: string, curline: string, curcol: string)
        # let us quit with a single keypress
        nnoremap q <Cmd>quitall!<CR>

        # preserve cursor position
        var cmd: string = printf('normal! %sGzt%sG%s|', topline, curline, curcol)
        timer_start(0, (_) => execute(cmd))
    enddef

In legacy Vimscript:

    command -bar -nargs=+ KittyScrollback call s:KittyScrollback(<f-args>)

    function s:KittyScrollback(topline, curline, curcol)
        " let us quit with a single keypress
        nnoremap <silent> q :<C-U>quitall!<CR>

        " preserve cursor position
        let cmd = printf('normal! %sGzt%sG%s|', a:topline, a:curline, a:curcol)
        call timer_start(0, {-> execute(cmd)})
    endfunction

Tested with kitty on [this commit](https://github.com/kovidgoyal/kitty/commit/f047678711e02316f8679d71628937ffa73f698b), and with Vim up to the patch [8.2.4306](https://github.com/vim/vim/releases/tag/v8.2.4306).

**Reply by lacygoill** [reply]
You might also keep using `less(1)` as the default pager, and in [its configuration file](https://github.com/lacygoill/config/blob/0b7d442c4ec6afeef8850ce816cdbc092ec80977/.config/less/keys#L182-L196), you could install this key binding:

    E noaction g|$vim -^M

From then, if you press `E` while in `less(1)`, the latter should pipe the text to Vim.

You need to compile the file with `lesskey(1)`.  I have [this autocmd](https://github.com/lacygoill/config/blob/0b7d442c4ec6afeef8850ce816cdbc092ec80977/.vim/vimrc#L8268-L8269) to do it automatically:

    autocmd BufWritePost ~/.config/less/keys printf('lesskey %s/.config/less/keys', $HOME)->system()

**Reply by jclsn** [reply]
@lacygoill Thanks, the solution is similar to the one posted by page-down. Unfortunately both don't give you the possibility to edit the current command. But well, Alacritty can't do that either

**Reply by trygveaa** [reply]
@jclsn: If you want to edit the current command in the shell it's more appropriate to use the vi mode of the shell. `bindkey -v` for zsh or `set -o vi` for bash.

**Reply by jclsn** [reply]
@trygveaa I don't understand how to use that. I am in zsh. There is a zsh-vi-mode plugin, but it breaks the normal history behavior

EDIT: Okay got it. This is actually fairly helpful. Wonder why there is a zsh-vi-mode plugin after all if this already existed.

**Reply by lacygoill** [reply]
> @lacygoill Unfortunately both don't give you the possibility to edit the current command.

In bash, the readline library provides the `edit-and-execute-command` command:

    edit-and-execute-command (C-x C-e)
           Invoke an editor on the current command line,  and  execute  the
           result as shell commands.  Bash attempts to invoke $VISUAL, $ED‐
           ITOR, and emacs as the editor, in that order.

By default it's bound to `CTRL-X CTRL-E`, but you could rebind it to whatever key sequence you want; e.g. `CTRL-X CTRL-X`:

    bind '"\C-x\C-X":  edit-and-execute-command'

In zsh, the zle editor provides the `edit-command-line` command:

    autoload -Uz edit-command-line
    zle -N edit-command-line
    bindkey '^X^E' edit-command-line

**Reply by jclsn** [reply]
@lacygoill The ``bindkey -v`` solution is much simpler. You just add it to your .zshrc and then you can invoke it by pressing escape

**Reply by lacygoill** [reply]
Sure, use what works best for you.  For me, what works best is `edit-command-line`.  It's especially useful for editing commands which are split on multiple lines.

**Reply by jclsn** [reply]
@lacygoill I have improved the KittyScrollBack function a bit, because I don't want to depend on Neovim for this. How can I make the window use the whole terminal space? Your solution only uses a small part.

```
scrollback_pager bash -c 'vim --not-a-term +"terminal ++curwin cat "<(cat) +"KittyScrollback INPUT_LINE_NUMBER CURSOR_LINE CURSOR_COLUMN"'
```
```
command -bar -nargs=+ KittyScrollback call s:KittyScrollback(<f-args>)

function s:KittyScrollback(topline, curline, curcol)
    " let us quit with a single keypress
    nnoremap <silent> q :<C-U>quitall!<CR>
    nnoremap <silent> i :<C-U>quitall!<CR>
    set laststatus=0
    set nonu
    set nornu
    set clipboard=autoselect,unnamedplus,exclude:cons\|linux

    " preserve cursor position
    let cmd = printf('normal! %sGzt%sG%s|', a:topline, a:curline, a:curcol)
    call timer_start(0, {-> execute(cmd)})
endfunction
```
Also, when the terminal output is too long, it doesn't work at all. It goes back directly to Kitty after scrolling through all the output.

**Reply by jclsn** [reply]
@lacygoill The cursor is still a bit off when the terminal is filled with text to the bottom and you use the KittyScrollback function. Any idea how to fix this?
